File: 25.reverse-nodes-in-k-group.java
/*
 * @lc app=leetcode id=25 lang=java
 *
 * [25] Reverse Nodes in k-Group
 */
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        
        if (head == null || head.next == null) return head;
        ListNode be4Head = new ListNode(0);
        be4Head.next = head;
        boolean init = false;
        ListNode be4Itr = be4Head, firstItemAtItr = be4Itr.next;
        ListNode prev = be4Head, cur = prev.next, nex = cur.next;

        while (firstItemAtItr != null) {
            int itr = k;
            outer:
            for (int dump = 1; dump <= 2; ++dump) {

                while (itr > 0) {
                    if (nex == null && itr > 1) {
                        nex = prev;
                        prev = null;
                        itr = k - itr + 1;
                        continue outer;
                    }
                    cur.next = prev;
                    prev = cur;
                    cur = nex;
                    if (nex != null) nex = nex.next;
                    itr--;
                }

                if (dump == 2) return be4Head.next;
                break;
            }

            be4Itr.next = prev;
            if (!init) {
                be4Head.next = prev;
                init = true;
            }
            firstItemAtItr.next = cur;
            be4Itr = firstItemAtItr;
            firstItemAtItr = cur;
            prev = be4Itr;
        }
        
        return be4Head.next;
    }
}

File: go/25.reverse-nodes-in-k-group.go
/*
 * @lc app=leetcode id=25 lang=golang
 *
 * [25] Reverse Nodes in k-Group
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseKGroup(head *ListNode, k int) *ListNode {
    if head == nil || k <= 1 {
		return head
	}
	ahead := &ListNode{0, head}
	prev, cur, next := ahead, head, head.Next
	OUTER:
	for {
		tmp := cur
		for i := 0; i < k; i++ {
			if tmp == nil {
				break OUTER
			}
			tmp = tmp.Next
		}
		firstPrev := prev
		for i := 0; i < k; i++ {
			cur.Next = prev
			prev, cur = cur, next
			if next != nil {
				next = next.Next
			}
		}
		tmp = prev
		prev = firstPrev.Next
		firstPrev.Next.Next = cur
		firstPrev.Next = tmp
	}
	return ahead.Next
}
// @lc code=end

